File: LeetCode/cpp/lc_0977.cpp
/**
 * @file     lc_0977.cpp
 * @brief    https://leetcode-cn.com/problems/squares-of-a-sorted-array/
 * @author   YongDu
 * @date     2021-09-05
 */
class Solution {
  public:
    vector<int> sortedSquares(vector<int>& nums) {
        int              len = nums.size();
        std::vector<int> result(len);
        int              k = len - 1;
        for (int i = 0, j = len - 1; i <= j;) {
            if (std::pow(nums[i], 2) >= std::pow(nums[j], 2)) {
                result[k--] = std::pow(nums[i++], 2);
            } else {
                result[k--] = std::pow(nums[j--], 2);
            }
        }
        return result;
    }
};

File: LeetCode/python/lc_0977.py
"""
 * @file     lc_0977.py
 * @brief    https://leetcode-cn.com/problems/squares-of-a-sorted-array/
 * @author   YongDu
 * @date     2021-09-09
"""


class Solution:
    def sortedSquares(self, nums: List[int]) -> List[int]:
        l, r, k = 0, len(nums) - 1, len(nums) - 1
        res = [-1] * len(nums)
        while l <= r:
            lpro = nums[l] ** 2
            rpro = nums[r] ** 2
            if lpro > rpro:
                res[k] = lpro
                l += 1
            else:
                res[k] = rpro
                r -= 1
            k -= 1
        return res
